### PR TITLE
AC-400: add concepts data providers, add visit form

### PR DESCRIPTION
**Pull request by Gelassen**
## Description of what I changed
The task was to make a data provider for concepts, that can operate both REST and cache sources. Feature for concepts that was cached already implemented few years ago and the same API was used for web request (SIC! apparently that concepts feature wasn't fully correct despite it passed code review, see details in ticket comments).

I make data provider for REST call and combine it with cache. I make a visit form as PoC of concepts autopopulate feature. I extend business logic of visit form to populate with all necessary data. Altought I faced with issue to find out way to submit entered data. It should be done to make visit note business valuable.  

Also I rename existing data providers from API postfix to Repository postfix as it more relevant to their current status and common naming convention. 
I introduced several business layers according to clean architecture principles in addition to MVP as current state of project needs it. 

## Issue I worked on
JIRA Issue: https://issues.openmrs.org/browse/AC-400

## Checklist: I completed these to help reviewers :)
- [x] My pull request only contains **ONE single commit**

- [ ] I have **added tests** to cover my changes. (If you refactored
existing code that was well tested you do not have to add tests)
(I noticed previous contributors stops cover changes with tests. Does it link to some extra work should be done to mock some classes?)

- [ ] All new and existing **tests passed**.
(Half of the tests fails. By looking through errors like this "Reason: cannot find android.support.v4.util.LruCache" seems they reffred to work of previous contributor. I think it is a good manners to fix them anyway, but let me do PR first to submit existing work to community as I contribute spare time to this project which I might lack of it at any time)

- [x] My pull request is **based on the latest changes** of the master branch.

### CODE REVIEW

**Comment by f4ww4z** [comment]
@Gelassen this merge conflict is what I mean when I said to wait for 1 PR to be merged first before moving to the next one. Kindly fix it first.

**Comment by Gelassen** [comment]
@f4ww4z thank you.  I fixed the merge conflict. 

There is an issue with tests. I addressed cause when I made PR at the firtst time, but there is something else in travis-ci console.

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/providermanagerdashboard/ProviderManagerDashboardPresenter.java [hunk 1]
```diff
@@ -19,7 +19,7 @@
 import org.jetbrains.annotations.NotNull;
 import org.openmrs.mobile.activities.BasePresenter;
 import org.openmrs.mobile.api.CustomApiCallback;
-import org.openmrs.mobile.api.retrofit.ProviderRepository;
+import org.openmrs.mobile.api.repository.ProviderRepository;
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/providermanagerdashboard/ProviderManagerDashboardPresenter.java]: This is fixed in #614 .

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/patientdashboard/visits/PatientDashboardVisitsPresenter.java [hunk 2]
```diff
@@ -31,13 +31,15 @@
     private PatientDashboardContract.ViewPatientVisits mPatientVisitsView;
     private VisitDAO visitDAO;
     private VisitRepository visitRepository;
+    private VisitRepository visitApi;
```
Comment (f4ww4z) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/patientdashboard/visits/PatientDashboardVisitsPresenter.java]: Why have 2 `VisitRepository` instances?

**Comment by f4ww4z** [comment]
@Gelassen Are you still working on this? See the failing CI build.

**Comment by f4ww4z** [comment]
Closing due to inactivity. @Gelassen Ping me when you resume working on it.

**Pull request closed without merging**